File: lib/cal_path.c
//
//  General-purpose Photovoltaic Device Model gpvdm.com- a drift diffusion
//  base/Shockley-Read-Hall model for 1st, 2nd and 3rd generation solarcells.
// 
//  Copyright (C) 2012 Roderick C. I. MacKenzie r.c.i.mackenzie at googlemail.com
//
//	https://www.gpvdm.com
//	Room B86 Coates, University Park, Nottingham, NG7 2RD, UK
//
//
// This program is free software; you can redistribute it and/or modify it
// under the terms and conditions of the GNU General Public License,
// version 2, as published by the Free Software Foundation.
//
// This program is distributed in the hope it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
// more details.

/** @file cal_path.c
	@brief Calculate the path of where stuff is, and if it can't find a file look harder.  Win/Linux.
*/


#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include "cal_path.h"
#include "util.h"
#include "inp.h"
#include <log.h>

#include <unistd.h>
#include <dirent.h>
#include <fcntl.h>
#include <stdarg.h>
#include <const.h>


#include <limits.h>

int get_delta_path(struct simulation *sim,char *out, char *root,char *file_name)
{
int root_len=strlen(root);
int file_name_len=strlen(file_name);
if (root_len>file_name_len)
{
	strcpy(out,file_name);
	return -1;
}

if (root_len==0)
{
	strcpy(out,file_name);
	return 0;
}
if (strcmp_begin(file_name,root)==0)
{
	strcpy(out,file_name+root_len+1);
	return 0;
}

return -1;
}

int find_dll(struct simulation *sim, char *lib_path,char *lib_name)
{
char full_name[PATHLEN];
char temp[PATHLEN];
sprintf(full_name,"%s.so",lib_name);

join_path(2,lib_path,get_plugins_path(sim),full_name);
if (isfile(lib_path)==0)
{
	return 0;
}
	
struct dirent *next_file;
DIR *theFolder;

theFolder = opendir(get_plugins_path(sim));
if (theFolder!=NULL)
{
	while((next_file=readdir(theFolder))!=NULL)
	{
		split_dot(temp, next_file->d_name);
		if (strcmp(lib_name,temp)==0)
		{
			join_path(2,lib_path,get_plugins_path(sim),next_file->d_name);
			if (isfile(lib_path)==0)
			{
				closedir (theFolder);
				return 0;
			}
				
		}
	}

closedir (theFolder);

}

ewe(sim,"I can't find the dll %s,\n",lib_name);

return -1;
}

void set_path(struct simulation *sim,char *out, char *name)
{
char cwd[PATHLEN];
char temp[PATHLEN];

	if (getcwd(cwd,PATHLEN)==NULL)
	{
		ewe(sim,"IO error\n");
	}

	join_path(2,temp,cwd,name);

	if (isdir(temp)==0)
	{
		strcpy(out,temp);
		return;
	}

	join_path(2,temp,sim->exe_path,name);
	printf("search %s",sim->exe_path);
	if (isdir(temp)==0)
	{
		strcpy(out,temp);
		return;
	}
	
	join_path(2,temp,"/usr/lib/gpvdm/",name);
	if (isdir(temp)==0)
	{
		strcpy(out,temp);
		return;
	}

	join_path(2,temp,"/usr/lib64/gpvdm/",name);
	if (isdir(temp)==0)
	{
		strcpy(out,temp);
		return;
	}
	
	join_path(2,temp,"/usr/share/gpvdm/",name);
	if (isdir(temp)==0)
	{
		strcpy(out,temp);
		return;
	}
 
	//Ubuntu
	join_path(2,temp,"/usr/lib/x86_64-linux-gnu/gpvdm/",name);
	if (isdir(temp)==0)
	{
		strcpy(out,temp);
		return;
	}

	join_path(2,temp,sim->share_path,name);
	if (isdir(temp)==0)
	{
		strcpy(out,temp);
		return;
	}
	
	ewe(sim,"I can't find the %s\n",name);

}

void get_file_name_from_path(char *out,char *in)
{
	int i=0;

	strcpy(out,in);

	if (strlen(in)==0)
	{
		return;
	}

	for (i=strlen(in)-1;i>=0;i--)
	{
		if ((in[i]=='\\') || (in[i]=='/'))
		{
			i++;
			strcpy(out,(char*)(in+i));
			return;
		}
	}
	
}

int is_dir_in_path(char *long_path, char* search_dir)
{
	if( strstr(long_path, search_dir) != NULL)
	{
		return 0;
	}else
	{
		return -1;
	}
return -1;
}

void get_nth_dir_name_from_path(char *out,char *in,int n)
{
	int i=0;
	int ii=0;
	int pos=0;
	int start=0;
	int stop=0;
	int count=0;
	strcpy(out,in);

	if (strlen(in)==0)
	{
		return;
	}

	for (i=0;i<strlen(in);i++)
	{
		if ((in[i]=='\\') || (in[i]=='/') || (i==strlen(in)-1))
		{
			if (i!=0)
			{
				stop=i;
				if (i==strlen(in)-1)
				{
					stop++;
				}


				if (count==n)
				{

					for (ii=start;ii<stop;ii++)
					{
						out[pos]=in[ii];
						pos++;
					}
					out[pos]=0;
					return;
				}
				start=stop+1;
				count++;
			}else
			{
				start=i+1;		// move it past the first / in a unix string
			}
		}
	}
	
}
void cal_path(struct simulation *sim)
{
char cwd[PATHLEN];
char temp[PATHLEN];

strcpy(cwd,"");
strcpy(temp,"");

strcpy(sim->share_path,"nopath");

strcpy(sim->plugins_path,"");
strcpy(sim->lang_path,"");
strcpy(sim->input_path,"");
strcpy(sim->output_path,"");

memset(temp, 0, PATHLEN * sizeof(char));
int len = readlink("/proc/self/exe", temp, PATHLEN);
if (len == -1)
{
	ewe(sim,"IO error\n");
}


get_dir_name_from_path(sim->exe_path, temp);

if (isfile("configure.ac")==0)
{
	strcpy(sim->share_path,cwd);
	//printf_log(sim,"share path: %s\n",sim->share_path);
}else
if (isfile("ver.py")==0)
{
	path_up_level(temp, cwd);
	strcpy(sim->share_path,temp);
	//printf_log(sim,"share path: %s\n",sim->share_path);
}else
{
	strcpy(sim->share_path,"/usr/lib64/gpvdm/");
}

if (getcwd(cwd,PATHLEN)==NULL)
{
	ewe(sim,"IO error\n");
}

strcpy(sim->root_simulation_path,cwd);
strcpy(sim->output_path,cwd);
strcpy(sim->input_path,cwd);
set_path(sim,sim->plugins_path, "plugins");
set_path(sim,sim->lang_path, "lang");
set_path(sim,sim->materials_path, "materials");
set_path(sim,sim->spectra_path, "spectra");

}

char *get_spectra_path(struct simulation *sim)
{
return sim->spectra_path;
}


char *get_materials_path(struct simulation *sim)
{
return sim->materials_path;
}

char *get_plugins_path(struct simulation *sim)
{
return sim->plugins_path;
}

char *get_lang_path(struct simulation *sim)
{
return sim->lang_path;
}

char *get_input_path(struct simulation *sim)
{
return sim->input_path;
}

char *get_output_path(struct simulation *sim)
{
return sim->output_path;
}

void set_output_path(struct simulation *sim,char *in)
{
strcpy(sim->output_path,in);
}

void set_input_path(struct simulation *sim,char *in)
{
strcpy(sim->input_path,in);
}


void join_path(int max, ...)
{
	max=max+1;
	char temp[PATHLEN];
	strcpy(temp,"");
	va_list arguments;
	int i;
	va_start ( arguments, max );
	char *ret=va_arg ( arguments, char * );
	strcpy(ret,"");
	for (i = 1; i < max; i++ )
	{
		if ((i!=1)&&(strcmp(temp,"")!=0))
		{
			strcat(ret,"/");
		}
		strcpy(temp,va_arg ( arguments, char * ));
		strcat(ret,temp);
	}
	va_end ( arguments );                  // Cleans up the list

	return;
}


/**Make sure the slashes go the right way in a string for which ever OS we are on.
@param path path to check
*/
void assert_platform_path(char * path)
{
	int i=0;
	char temp[PATHLEN];
	strcpy(temp,"");
	int max=strlen(path);
	for (i=0;i<max;i++)
	{
		if ((path[i]=='\\')||(path[i]=='/'))
		{
			strcat(temp,"/");
		}else
		{
			temp[i]=path[i];
			temp[i+1]=0;
		}


	}

	strcpy(path,temp);

	return;
}

File: gui/cal_path.py
#    General-purpose Photovoltaic Device Model - a drift diffusion base/Shockley-Read-Hall
#    model for 1st, 2nd and 3rd generation solar cells.
#    Copyright (C) 2012 Roderick C. I. MacKenzie r.c.i.mackenzie at googlemail.com
#
#	https://www.gpvdm.com
#	Room B86 Coates, University Park, Nottingham, NG7 2RD, UK
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License v2.0, as published by
#    the Free Software Foundation.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License along
#    with this program; if not, write to the Free Software Foundation, Inc.,
#    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


import sys
import os
#import shutil
from win_lin import running_on_linux
from gui_enable import gui_get
	
if running_on_linux()==False:
	import winreg

materials_path=None
plugins_path=None
exe_command=None
share_path=None
device_lib_path=None
bin_path=None
lib_path=None
image_path=None
css_path=None
flag_path=None
lang_path=None
inp_file_path=None
src_path=None
spectra_path=None
sim_path=None
materials_base_path=None
html_path=None

def subtract_paths(root,b_in):
	a=root.replace("/","\\")
	b=b_in.replace("/","\\")
	a=a.split("\\")
	b=b.split("\\")
	a_len=len(a)
	b_len=len(b)
	m=a_len
	if b_len<m:
		m=b_len
	pos=0

	for i in range(0,m):
		if a[i]!=b[i]:
			break
		pos=pos+1

	ret=[]
	for i in range(pos,b_len):
		ret.append(b[i])

	return "/".join(ret)

def to_native_path(path):
	ret=path
	if running_on_linux()==False:
		ret=ret.replace("/","\\")
		ret=ret.lower()
	return ret

def remove_cwdfrompath(path):
	tmp=path
	if tmp.startswith(os.getcwd()):
		tmp=tmp[len(os.getcwd())+1:]
	return tmp

def remove_simpathfrompath(path):
	tmp=path
	if tmp.startswith(get_sim_path()):
		tmp=tmp[len(get_sim_path())+1:]
	return tmp

def join_path(one,two):
	output_file=os.path.join(one,two)

	if two[0]=='/':
		if one!="" :
			output_file=os.path.join(one,two[1:])

	return output_file


def cal_share_path():
	global share_path

	if os.path.isfile("configure.ac"):
		share_path=os.getcwd()
		return

	if os.path.isfile("ver.py"):
		share_path=os.path.abspath(os.path.join(os.getcwd(), os.pardir))
		return

	if running_on_linux()==False:
		try:
			registry_key = winreg.OpenKey(winreg.HKEY_CURRENT_USER, "Software\\gpvdm", 0, winreg.KEY_READ)
			value, regtype = winreg.QueryValueEx(registry_key, "installpath")
			winreg.CloseKey(registry_key)
			share_path=value
		except WindowsError:
			if os.path.isfile(os.path.join(os.getcwd(),"gpvdm_core.exe")):
				share_path=os.getcwd()
			else:
				share_path="c:\\gpvdm"

			print("No registry key found using default",share_path)
	else:
		if os.path.isdir("/usr/lib64/gpvdm"):
			share_path="/usr/lib64/gpvdm/"
		elif os.path.isdir("/usr/lib/gpvdm"):
			share_path="/usr/lib/gpvdm/"
		else:
			share_path=os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
			print("I don't know where the shared files are assuming ",share_path)

def search_known_paths(file_or_dir_to_find,ext,key_file):
	global share_path
	global bin_path
	#check cwd
	paths=[]
	for ex in ext:
		paths.append(os.path.join(os.getcwd(),file_or_dir_to_find)+ex)
		paths.append(os.path.join(os.path.abspath(os.path.join(os.getcwd(), os.pardir)),file_or_dir_to_find)+ex)
		paths.append(os.path.join(share_path,file_or_dir_to_find)+ex)
		paths.append(os.path.join(bin_path,file_or_dir_to_find)+ex)
		paths.append(os.path.join(get_sim_path(),file_or_dir_to_find)+ex)
		if running_on_linux()==True:
			paths.append(os.path.join("/usr/share/gpvdm/",file_or_dir_to_find)+ex)
			paths.append(os.path.join("/usr/local/bin/",file_or_dir_to_find)+ex)
			paths.append(os.path.join("/usr/bin/",file_or_dir_to_find)+ex)

	for item in paths:
		if key_file==None:
			if os.path.isdir(item) or os.path.isfile(item):
				#print "found",item
				return to_native_path(item)
		else:
			if os.path.isfile(os.path.join(item,key_file)):
				return to_native_path(item)
	#print "Can't find",file_or_dir_to_find, "setting it to",paths[0]
	return paths[2]

def cal_bin_path():
	global bin_path
	if running_on_linux()==True:
			bin_path="/bin/"
	else:
			bin_path=share_path

def test_arg_for_sim_file():
	if len(sys.argv)>1:
		f=os.path.realpath(sys.argv[1])
		if os.path.isfile(f)==True and f.endswith("sim.gpvdm"):
			return os.path.dirname(f)
		elif os.path.isdir(f)==True and os.path.isfile(os.path.join(f,"sim.gpvdm"))==True:
			return f
	return False

def calculate_paths_init():
	set_sim_path(os.getcwd())
	path=test_arg_for_sim_file()
	if path!=False:
		set_sim_path(path)

	cal_share_path()
	cal_bin_path()

def get_icon_path(name,size=-1):
	global image_path
	if name.endswith(".png"):
		name=name[:-4]

	if size==-1:
		path=os.path.join(image_path,"64x64",name+".png")
		return path

	return os.path.join(image_path,str(size)+"x"+str(size),name+".png")


def calculate_paths():
	global share_path
	global lib_path
	global exe_command
	global device_lib_path
	global materials_path
	global image_path
	global css_path
	global flag_path	
	global plugins_path
	global lang_path
	global inp_file_path
	global src_path
	global ui_path
	global spectra_path
	global materials_base_path
	global html_path

	materials_path=os.path.join(get_sim_path(),"materials")
	if os.path.isdir(materials_path)==False:
		materials_path=search_known_paths("materials",[""],None)

	materials_base_path=search_known_paths("materials",[""],None)
	device_lib_path=search_known_paths("device_lib",[""],None)
	plugins_path=search_known_paths("plugins",[""],None)
	image_path=search_known_paths("images",[""],"image.jpg")
	css_path=search_known_paths("css",[""],"style.css")
	flag_path=search_known_paths("flags",[""],"gb.png")
	lang_path=search_known_paths("lang",[""],None)
	exe_command=search_known_paths("gpvdm_core",["",".exe",".o"],None)
	inp_file_path=os.path.dirname(search_known_paths("base",[".gpvdm"],None))
	src_path=os.path.dirname(search_known_paths("Makefile",[".am"],None))
	ui_path=search_known_paths("ui",[""],None)
	spectra_path=search_known_paths("spectra",[""],None)

	html_path=search_known_paths("html",[""],"info0.html")

def get_share_path():
	global share_path
	return share_path

def get_src_path():
	global src_path
	return src_path

def get_spectra_path():
	global spectra_path
	return spectra_path

def get_html_path():
	global html_path
	return html_path

def get_materials_path():
	global materials_path
	return materials_path

def get_base_material_path():
	global materials_base_path
	return materials_base_path

def get_default_material_path():
	global materials_base_path
	return os.path.join(materials_base_path,"generic","default")

def get_base_spectra_path():
	global spectra_path
	return os.path.join(spectra_path,"sun")

def get_device_lib_path():
	global device_lib_path
	return device_lib_path

def get_bin_path():
	global bin_path
	return bin_path

def get_plugins_path():
	global plugins_path
	return plugins_path

def get_exe_path():
	global exe_command
	ret=os.path.dirname(exe_command)
	return ret

def get_exe_command():
	global exe_command
	return exe_command

def get_exe_name():
	global exe_command
	return os.path.basename(exe_command)

def get_inp_file_path():
	global inp_file_path
	return inp_file_path

def get_image_file_path():
	global image_path
	return image_path

def get_css_path():
	global css_path
	return css_path

def get_flag_file_path():
	global flag_path
	return flag_path

def get_lang_path():
	global lang_path
	return lang_path

def get_ui_path():
	global ui_path
	return ui_path

def set_sim_path(path):
	global sim_path
	sim_path=os.path.abspath(path)

def get_sim_path():
	global sim_path
	if sim_path==None:
		return os.getcwd()
	return sim_path

def get_exe_args():
	if gui_get()==True:
		return "--gui --html" #--english
	else:
		return ""

def find_materials():
	ret=[]
	mat_path=get_materials_path()
	for dirpath, dirnames, filenames in os.walk(mat_path):
		for filename in [f for f in filenames if f=="mat.inp"]:
			path=os.path.join(dirpath, filename)
			path=os.path.dirname(path)
			s=os.path.relpath(path, mat_path)
			s=s.replace("\\","/")
			ret.append(s)

	return ret

def find_light_source():
	ret=[]

	spectra_path=get_spectra_path()

	for dirpath, dirnames, filenames in os.walk(spectra_path):
		for filename in [f for f in filenames if f=="mat.inp"]:
			path=os.path.join(dirpath, filename)
			path=os.path.dirname(path)
			s=os.path.relpath(path, spectra_path)
			s=s.replace("\\","/")
			ret.append(s)

#	for file in glob.glob(os.path.join(path,"*.spectra")):
#		ret.append(os.path.splitext(os.path.basename(file))[0])

	return ret
